### PR TITLE
iOS - Sync 1.15.x - Uplift

### PR DESCRIPTION
**Pull request by Brandon-T**
- Uplift of https://github.com/brave/brave-core/pull/6722 for 1.15.x

- Adds support for iOS Sync chain with Brave Desktop.
- Adds support for iOS Chromium bookmarks so we can sync them with Desktop.
- Adds importing and exporting Bookmarks to HTML for iOS.
- Adds better compatibility with iOS and Chromium core in general.
- All interfaces that are native to iOS must be explicitly exported with OBJC_EXPORT
- Rewards database code runs on Chromium Threads instead of iOS DispatchQueues.

## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [ ] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [ ] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: ios/browser/api/bookmarks/importer/brave_bookmarks_importer.mm [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  self.cancelled = true;
+}
+
+- (void)importFromFile:(NSString *)filePath
+    topLevelFolderName:(NSString *)folderName
+       automaticImport:(bool)automaticImport
+          withListener:(void(^)(BraveBookmarksImporterState, NSArray<BraveImportedBookmark *> * _Nullable))listener {
+  
+  base::FilePath source_file_path =
+      base::FilePath::FromUTF8Unsafe([filePath UTF8String]);
```
Comment (bridiver) [inline ios/browser/api/bookmarks/importer/brave_bookmarks_importer.mm:85]: use `base::mac::NSStringToFilePath`

File: ios/browser/api/bookmarks/importer/brave_bookmarks_importer.mm [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+- (instancetype)init {
+    if ((self = [super init])) {
+      self.cancelled = false;
+    }
+    return self;
+}
+
+- (void)dealloc {
+  [self cancel];
+  self->import_thread_.reset();
```
Comment (bridiver) [inline ios/browser/api/bookmarks/importer/brave_bookmarks_importer.mm:72]: this isn't necessary

File: ios/browser/api/bookmarks/importer/brave_bookmarks_importer.mm [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+      }
+    } else {
+      listener(BraveBookmarksImporterStateCancelled, nullptr);
+    }
+  };
+  
+  // Create worker thread in which importer runs.
+  // In Chromium, this is created with `base::Thread("import_thread")`
+  if (!self->import_thread_) {
+    self->import_thread_ = base::CreateSequencedTaskRunner(
```
Comment (bridiver) [inline ios/browser/api/bookmarks/importer/brave_bookmarks_importer.mm:136]: this should be created in the constructor

File: build/rust/Cargo.lock [hunk 4]
```diff
@@ -30,7 +30,7 @@ dependencies = [
 [[package]]
 name = "adblock"
 version = "0.3.1"
-source = "git+https://github.com/brave/adblock-rust?rev=1f8dd4273ebc0339babb1cc3e38bd6f14d44e7e1#1f8dd4273ebc0339babb1cc3e38bd6f14d44e7e1"
+source = "git+https://github.com/brave/adblock-rust?rev=9032b8bb07bb71879120fc8228c0d4d1d75a0894#9032b8bb07bb71879120fc8228c0d4d1d75a0894"
```
Comment (bridiver) [inline build/rust/Cargo.lock:33]: was this change intentional?

File: chromium_src/components/sync_device_info/device_info.cc [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	  return "ios";
+	}
+
+	if (user_agent.find("android") != std::string::npos) {
+	  return "android";
+	}
+	return "unknown";
+  }
+
+  std::unique_ptr<base::DictionaryValue> DeviceInfo::ToValue() const {
```
Comment (darkdh) [inline chromium_src/components/sync_device_info/device_info.cc:26]: we can call `ToValue_ChromiumImpl()` to get the value and then set our own os string additionally.
So we don't have to duplicate logic

File: chromium_src/ios/chrome/browser/application_context_impl.mm [hunk 6]
```diff
@@ -0,0 +1,356 @@
+// Copyright 2014 The Chromium Authors. All rights reserved.
```
Comment (AlexeyBarabash) [inline chromium_src/ios/chrome/browser/application_context_impl.mm:1]: Why do we have two copyrights here? Is it some special case?

File: ios/browser/api/bookmarks/exporter/bookmark_html_writer.cc [hunk 7]
```diff
@@ -0,0 +1,631 @@
+#include "brave/ios/browser/api/bookmarks/exporter/bookmark_html_writer.h"
```
Comment (AlexeyBarabash) [inline ios/browser/api/bookmarks/exporter/bookmark_html_writer.cc:1]: nit: missing copyright header

File: ios/browser/api/bookmarks/exporter/bookmarks_encoder.cc [hunk 8]
```diff
@@ -0,0 +1,186 @@
+#include "brave/ios/browser/api/bookmarks/exporter/bookmarks_encoder.h"
```
Comment (AlexeyBarabash) [inline ios/browser/api/bookmarks/exporter/bookmarks_encoder.cc:1]: nit: also missing copyright header

File: ios/browser/api/sync/brave_sync_api.mm [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+        fprintf(stderr, "Sync Shut Down\n");
+    });
+  }
+  return self;
+}
+@end
+
+@interface BraveSyncAPI()
+{
+  std::unique_ptr<BraveSyncWorker> _worker;
```
Comment (AlexeyBarabash) [inline ios/browser/api/sync/brave_sync_api.mm:79]: Seeing influence of Android code :)

File: ios/browser/api/sync/brave_sync_worker.cc [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+    return false;
+  }
+  return true;
+}
+
+std::string BraveSyncWorker::GetSyncCodeFromHexSeed(const std::string& hex_code_seed) {
+  DCHECK(!hex_code_seed.empty());
+
+  std::vector<uint8_t> bytes;
+		std::string sync_code_words;
```
Comment (AlexeyBarabash) [inline ios/browser/api/sync/brave_sync_worker.cc:237]: nit: mad tabs

File: ios/browser/api/sync/brave_sync_worker.cc [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+		return sync_code_words;
+}
+
+bool BraveSyncWorker::IsFirstSetupComplete() {
+		syncer::SyncService* sync_service = GetSyncService();
+		return sync_service &&
+										sync_service->GetUserSettings()->IsFirstSetupComplete();
+}
+
+bool BraveSyncWorker::RemoveDevice(const std::string& device_guid) {
```
Comment (AlexeyBarabash) [inline ios/browser/api/sync/brave_sync_worker.cc:262]: this is nice and seems be useful when brave-core master will be able to remove other devices in chain :+1:

File: ios/browser/api/sync/brave_sync_worker.cc [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+    
+    tracker->DeleteDeviceInfo(device_itr->get(),
+                              base::BindOnce(&BraveSyncWorker::OnRemoteDeviceInfoDeleted,
+                                             weak_ptr_factory_.GetWeakPtr()));
+
+    return true;
+}
+
+bool BraveSyncWorker::ResetSync() {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
```
Comment (AlexeyBarabash) [inline ios/browser/api/sync/brave_sync_worker.cc:300]: Now the code around this place is changed in master around this place, and, if there are no iOS-specific obstacles, then it may be would be better to change so not to duplicate the code?

https://github.com/brave/brave-core/blob/f5a69e8d1735f2f724b04e2b5bf666cccc973b50/browser/android/brave_sync_worker.cc#L202

and 

https://github.com/brave/brave-core/blob/f5a69e8d1735f2f724b04e2b5bf666cccc973b50/components/brave_sync/profile_sync_service_helper.cc#L17

File: ios/browser/api/sync/brave_sync_worker.cc [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+}
+
+void BraveSyncWorker::OnSyncShutdown(syncer::SyncService* service) {
+  if (sync_service_observer_.IsObserving(service)) {
+    sync_service_observer_.Remove(service);
+  }
+}
+
+void BraveSyncWorker::OnRemoteDeviceInfoDeleted() {
+  VLOG(0) << __func__ << " local device successfully removed";
```
Comment (AlexeyBarabash) [inline ios/browser/api/sync/brave_sync_worker.cc:393]: The log message is about the local device, but the function is about remote device

File: ios/browser/api/sync/brave_sync_worker.cc [hunk 14]
```diff
@@ -0,0 +1,430 @@
...
+    sync_service_observer_.Add(sync_service);
+  }
+
+  setup_service->SetSyncEnabled(enabled);
+
+  if (enabled && !sync_service->GetUserSettings()->IsFirstSetupComplete()) {
+    setup_service->PrepareForFirstSyncSetup();
+    setup_service->SetFirstSetupComplete(
+        syncer::SyncFirstSetupCompleteSource::ADVANCED_FLOW_CONFIRM);
+  }
```
[the remark this answers is not in the document]
Comment (darkdh) [reply ios/browser/api/sync/brave_sync_worker.cc:170]: For ios, we also need to call `SyncSetupService::CommitSyncChanges` to remove `SetupInProgressHandle` that `PrepareForFirstSyncSetup` requested. When someone is holding `SetupInProgressHandle`, sync engine will pause and wait for it to be released.

**Review by N409704 — APPROVED** [review]

File: ios/browser/api/sync/brave_sync_worker.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (darkdh) [reply ios/browser/api/sync/brave_sync_worker.cc]: As we discussed on Slack, we just need to remove `PrepareForFirstSyncSetup` because on ios and android doesn't need to pause engine when setting up.

File: ios/browser/api/sync/brave_sync_worker.cc [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+		return sync_code_words;
+}
+
+bool BraveSyncWorker::IsFirstSetupComplete() {
+		syncer::SyncService* sync_service = GetSyncService();
+		return sync_service &&
+										sync_service->GetUserSettings()->IsFirstSetupComplete();
+}
+
+bool BraveSyncWorker::RemoveDevice(const std::string& device_guid) {
```
Comment (darkdh) [inline ios/browser/api/sync/brave_sync_worker.cc:263]: We need to hide deleting remote device until https://github.com/brave/brave-core/pull/6732 is merged. Otherwise, when remote device receives deleted record, it won't self reset. It is still on sync chain, the only thing changes is it is removed from device list.

File: build/rust/Cargo.lock (hunk 4 above)
Comment (Brandon-T) [reply build/rust/Cargo.lock:33]: No. reverted.

File: ios/browser/api/sync/brave_sync_worker.cc (hunk 12 above)
Comment (Brandon-T) [reply ios/browser/api/sync/brave_sync_worker.cc:301]: 1.15.x doesn't have such `brave-core/components/brave_sync/profile_sync_service_helper.cc` file. 
Only `master` has it.

**Comment by bsclifton** [comment]
Whoops - this ship sailed a while back 😂
